### PR TITLE
Emit bundler-friendly URL locators

### LINKED ISSUE #13571: Emit bundler-friendly URL locators
according to  https://twitter.com/RReverser/status/1365342246653018115 ,wasm-bindgen already support emit code import wasm using import, which is very helpful for bundler, I wonder whether emscripten could support the same feature?

### PR DESCRIPTION
**Pull request by RReverser**
This PR changes loading of main Wasm binary as well as helper Worker used by PThread integration to use a URL expression that can be both used directly by browsers as well as statically detected by bundlers like Webpack.

The main caveats are:
1. Emscripten is very configurable, so some of the new conditions might look odd but are necessary to keep backward compatibility and allow overriding bundler-friendly URL with a custom one during runtime.
2. None of Closure, our fork of Terser, or even `eval` (which is used by Emscripten's JS library preprocessing) support `import.meta` expressions without more work. While Closure seems to have _just_ implemented such support, and it wouldn't be too hard to add it to our Terser too, `eval` usage would still require a string replacement before execution (or complete revamp).
  To keep implementation simple, for now I went with just string replacement that covers all tools - this way, we replace `import.meta` -> `EMSCRIPTEN$IMPORT$META` only once when JS is added before any of this tooling is executed, and then replace back after everything is done right before the final emit.
  We might want to revisit this in future, but for now this works well and covers all the tooling incompatibilities together.
3. This won't work in Node.js, since it's not compatible with `EXPORT_ES6` in general yet.
4. I've only updated places for loading main Wasm binary and PThread code. This should cover majority of use-cases, but other external files like side modules, proxy-to-pthread, proxy-to-worker, external memory loading etc. are not covered by this PR and need to be updated separately if someone wants them to work with bundlers out of the box too.

Fixes #13571.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -1206,7 +1206,8 @@ def phase_setup(state):
   else:
     target = 'a.out.js'
 
-  settings.TARGET_BASENAME = unsuffixed_basename(target)
+  settings.TARGET_BASENAME_WITH_EXT = os.path.basename(target)
```
Comment (sbc100) [inline emcc.py:1209]: Does your code assume this file is main the javascript file?  Which case I would call this something more like `TARGET_JS_FILE` and delay assigning it until later down when we actual calculate the JS filename (see `js_target = `..  in `phase_final_emitting`..  I would just set it then to the same value is `js_target`).

File: emcc.py [hunk 2]
```diff
@@ -2534,6 +2535,16 @@ def phase_final_emitting(options, target, wasm_target, memfile):
     shared.JS.handle_license(final_js)
     shared.run_process([shared.PYTHON, shared.path_from_root('tools', 'hacky_postprocess_around_closure_limitations.py'), final_js])
 
+  # Unmangle previously mangled `import.meta` references in both main code and libraries.
+  # See also: `preprocess` in parseTools.js.
+  if settings.EXPORT_ES6 and settings.USE_ES6_IMPORT_META:
+    with open(final_js, 'r+') as f:
+      src = f.read()
+      src = src.replace('EMSCRIPTEN$IMPORT$META', 'import.meta')
```
Comment (sbc100) [inline emcc.py:2543]: if you use `do_replace` it will assert that the pattern was found.

File: emcc.py [hunk 3]
```diff
@@ -2534,6 +2535,16 @@ def phase_final_emitting(options, target, wasm_target, memfile):
...
+      f.seek(0)
+      f.write(src)
+      f.truncate()
```
Comment (sbc100) [inline emcc.py:2546]: Almost everything else we do to the JS is not directly in-place like this but instead creates a new temporary file which allows a debug build to see the complete history of the JS file as it processed (the sets final_js to point to the new file).  Maybe this is overkill, and it would be nice to create better, more centralized way of doing this.. but maybe better to stick with the existing convention here?

File: src/worker.js [hunk 4]
```diff
@@ -8,13 +8,53 @@
 // This is the entry point file that is loaded first by each Web Worker
 // that executes pthreads on the Emscripten application.
 
+'use strict';
```
Comment (sbc100) [inline src/worker.js:11]: I thought code that ran in workers (like modules) was implicitly running in strict mode.  I guess I got my wires cross somewhere?

File: src/worker.js [hunk 5]
```diff
@@ -8,13 +8,53 @@
 // This is the entry point file that is loaded first by each Web Worker
 // that executes pthreads on the Emscripten application.
```
Comment (sbc100) [inline src/worker.js:9]: IIUC all the changes to this file (or most of them anyway) are part of making worker run in strict mode?   Can split this into its own PR?

File: emcc.py (hunk 1 above)
Comment (RReverser) [reply emcc.py:1209]: > Does your code assume this file is main the javascript file?

Yes, but it needs to be relative so that it's usable in `import` expression, not the full path. If anything, I'd rename `TARGET_BASENAME` instead as it seems misleading - `basename` normally means "just the filename" (with extension included), whereas in Emscripten `TARGET_BASENAME` has extension already stripped.

> and delay assigning it until later down when we actual calculate the JS filename

Hm, is it ever going to be different than just basename?

File: emcc.py (hunk 2 above)
Comment (RReverser) [reply emcc.py:2543]: Ah nice, TIL. I'm trying to think if there are situations where code wouldn't use `import.meta` and we wouldn't want such assertions... maybe when using asm.js output?

File: emcc.py (hunk 3 above)
Comment (RReverser) [reply emcc.py:2546]: SGTM, I don't mind. I'm not entirely sure I understand the mechanism - aside from renaming files, do I need to also call `save_intermediate`? What does it do?

File: src/worker.js (hunk 4 above)
Comment (RReverser) [reply src/worker.js:11]: No, it's not. By default Workers are just regular scripts, although there is a recent feature where you can do `new Worker(..., { type: 'module' })` and then it will run in a strict module context.

File: src/worker.js (hunk 5 above)
Comment (RReverser) [reply src/worker.js:9]: Most of them, yeah. I could split them out, but this PR would still depend on them, as otherwise output is invalid.

**Comment by RReverser** [comment]
The failing test is interesting btw - I saw it failing locally, then reproduced it failing even without my changes, yet I see that it worked on CI on the last commit on main branch... I guess I'll need to look deeper :/

File: emcc.py (hunk 2 above)
Comment (RReverser) [reply emcc.py:2543]: Yeah, just checked, in case of asm.js there is no `import.meta` instances and correspondingly `.replace` won't find anything. There are probably more scenarios where this is true, so for simplicity I'll just keep `.replace` here.

File: emcc.py (hunk 1 above)
Comment (RReverser) [reply emcc.py:1209]: > so assuming that TARGET is JavaScript at this point is not correct.

Ohh true, I thought this was JS already. Good point, I'll fix that.

File: emcc.py (hunk 3 above)
Comment (sbc100) [reply emcc.py:2546]: That seems to be the pattern so far yes... as I say it would be nice to refactor/encapsulate that at some point.

File: src/worker.js (hunk 5 above)
Comment (RReverser) [reply src/worker.js:9]: Aside from code having to be carefully updated - no.

File: emcc.py (hunk 1 above)
Comment (RReverser) [reply emcc.py]: > and delay assigning it until later down when we actual calculate the JS filename (see js_target = .. in phase_final_emitting.. I would just set it then to the same value is js_target).

Hm, I can't actually delay it, because by that time all the macro has been already expanded to their values. Instead, I moved those `js_target ...` assignments to this phase, and reused computed value in the `phase_final_emitting`. I hope this is fine - it seems to work.

Btw, `EXPORT_ES6` is currently not compatible with `.html` outputs anyway (due to #7950 + #10086), and for `.wasm` outputs it's a no-op, so this might be not that big of a deal today, but worth fixing anyway.

File: src/worker.js (hunk 5 above)
Comment (RReverser) [reply src/worker.js:9]: Extracted here: https://github.com/emscripten-core/emscripten/pull/14137

File: emcc.py [hunk 6]
```diff
@@ -1206,6 +1206,12 @@ def phase_setup(state):
   else:
     target = 'a.out.js'
 
+  if options.oformat in (OFormat.JS, OFormat.MJS):
+    js_target = target
+  else:
+    js_target = get_secondary_target(target, '.js')
+  settings.TARGET_JS_NAME = js_target
+
   settings.TARGET_BASENAME = unsuffixed_basename(target)
```
[the remark this answers is not in the document]
Comment (RReverser) [reply emcc.py:1215]: It feels a bit awkward to split it up to a different place than `TARGET_BASENAME` given that both are semantically similar and are used in the same fashion (in templates) IMO.

re:`js_target` yeah I could skip it too, but felt it's more readable than using the long `settings.*` name in both branches.

[the remark this answers is not in the document]
Comment (sbc100) [reply emcc.py:1215]: I was suggesting you also move `settings.TARGET_BASENAME` since its also not needed until post-link.

I have a much bigger refactor almost ready to land so I can take care of moving this there if you prefer:
https://github.com/emscripten-core/emscripten/pull/14099

(One of us is going to have a conflict I think :)

[the remark this answers is not in the document]
Comment (RReverser) [reply emcc.py:1215]: Ah, sure, I can move both. Yeah, I guess a conflict is unavoidable, but my changes are relatively small, so I'm not too worried in this case :)

[the remark this answers is not in the document]
Comment (RReverser) [reply emcc.py]: Moved.

File: emcc.py [hunk 7]
```diff
@@ -2534,6 +2540,15 @@ def phase_final_emitting(options, target, wasm_target, memfile):
     shared.JS.handle_license(final_js)
     shared.run_process([shared.PYTHON, shared.path_from_root('tools', 'hacky_postprocess_around_closure_limitations.py'), final_js])
 
+  # Unmangle previously mangled `import.meta` references in both main code and libraries.
+  # See also: `preprocess` in parseTools.js.
+  if settings.EXPORT_ES6 and settings.USE_ES6_IMPORT_META:
+    src = open(final_js).read()
+    final_js += '.esmeta.js'
+    with open(final_js, 'w') as f:
+      f.write(src.replace('EMSCRIPTEN$IMPORT$META', 'import.meta'))
```
[the remark this answers is not in the document]
Comment (RReverser) [reply emcc.py:2549]: Well `import.meta` can generally appear 0 or more times in code depending on combination of options. If Wasm binary is loaded, that's one, if there is asm.js, that's zero. If PThreads are used, that's +1 to whatever was in previous step (so now 1 or 2).

If we add support for bundler-friendly URLs to other files in the future, there will be even more.

Generally we shouldn't rely on it appearing any specific number of times, so just `.replace` is most natural.

File: emcc.py [hunk 8]
```diff
@@ -2399,6 +2397,14 @@ def phase_post_link(options, in_wasm, wasm_target, target):
   if options.oformat != OFormat.WASM:
     final_js = in_temp(target_basename + '.js')
 
+  settings.TARGET_BASENAME = unsuffixed_basename(target)
+
+  if options.oformat in (OFormat.JS, OFormat.MJS):
+    js_target = target
+  else:
+    js_target = get_secondary_target(target, '.js')
+  settings.TARGET_JS_NAME = js_target
```
[the remark this answers is not in the document]
Comment (RReverser) [reply emcc.py:2406]: Idk, I can remove it if you insist, I'm just finding it a little easier to keep the long setting name in one place.

[the remark this answers is not in the document]
Comment (RReverser) [reply emcc.py]: Updated.

File: emcc.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sbc100) [reply emcc.py:2548]: Sure sgtm

Sorry for nitpicking.  Good to go.

**Comment by RReverser** [comment]
> PR description could use an update now that (3) landed separately. Otherwise feel free to land.

Updated!
